macos/ds_store.ksy

```diff
@@ -0,0 +1,193 @@
+meta:
+  id: ds_store
+  title: macOS '.DS_Store' format
+  license: CC-BY-SA-4.0
+  ks-version: 0.8
+  encoding: UTF-8
+  endian: be
+doc: |
+  Apple macOS '.DS_Store' file format.
+doc-ref: |
+  https://en.wikipedia.org/wiki/.DS_Store
+  https://metacpan.org/pod/distribution/Mac-Finder-DSStore/DSStoreFormat.pod
+  https://0day.work/parsing-the-ds_store-file-format
+seq:
+  - id: alignment_header
+    contents: [0x00, 0x00, 0x00, 0x01]
+  - id: buddy_allocator_header
+    type: buddy_allocator_header
+instances:
+  buddy_allocator_body:
+    pos: buddy_allocator_header.ofs_bookkeeping_info_block + 4
+    size: buddy_allocator_header.len_bookkeeping_info_block
+    type: buddy_allocator_body
+types:
+  buddy_allocator_header:
+    seq:
+      - id: magic
+        contents: ["Bud1"]
+        doc: Magic number 'Bud1'
+      - id: ofs_bookkeeping_info_block
+        type: u4
+      - id: len_bookkeeping_info_block
+        type: u4
+      - id: copy_ofs_bookkeeping_info_block
+        type: u4
+        doc: Needs to match 'offset_bookkeeping_info_block'
+      - size: 16
+        doc: Unused block
+  buddy_allocator_body:
+    seq:
+      - id: num_blocks
+        type: u4
+        doc: Number of blocks in the allocated-blocks list
+      - size: 4
+        doc: Unknown field
+      - id: block_addresses
+        type: u4
```

Comment:
```yaml
types:
  block_descriptor:
    seq:
       offset:
          type: u27
       size:
          type: u5
```

?

Reply:
Right. I didn't get that you meant adding a new type for the block addresses. I like this idea. However, it is a bit more complex than in your example. `offset_raw` needs to be a `u4` with the least-significant 5 bits set to 0 and the size needs to be equal to `2 ^ size`.

Consequently, it can be written like below:

```
  block_descriptor:
    seq:
      - id: address_raw
        type: u4
    instances:
      offset:
        value: (address_raw & ~0x1f) + 4
      size:
        value: 1 << address_raw & 0x1f
```

And then:

```
instances:
  master_block:
    pos: _parent.block_addresses[_parent.directory_entries[idx].block_id].offset
    size: _parent.block_addresses[_parent.directory_entries[idx].block_id].size
    type: master_block
```

Nevertheless, and despite that it works perfectly well, the offset and size values displayed in the Web IDE are incorrect:

![Screenshot from 2019-04-09 01-02-31](https://user-images.githubusercontent.com/9932586/55764411-4062f500-5a63-11e9-95a2-36c940d1a6c7.png)

Any idea why? I pushed these changes to https://github.com/SkypLabs/kaitai_struct_formats/tree/macos_ds_store_block_descriptor for your information.

Reply:
>bit more complex than in your example. offset_raw needs to be a u4 with the least-significant 5 bits set to 0 and the size needs to be equal to 2 ^ size.

You are right.

>despite that it works perfectly well, the offset and size values displayed in the Web IDE are incorrect

I have stopped using WebIDE because it has some bugs fixed by noone. In my case (another file format) it also visualizes incorrectly, so I had to write a convertor of parsed objects into tag files for wxHexEditor (IMHO wXHexEditor needs support for nested tags desperately).

Reply:
Ok I see. I will create an issue in the Web IDE's Git repository to report the bug later on.